### PR TITLE
Modify controller for backup operation

### PR DESCRIPTION
**Pull request by aerokite**

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -4,38 +4,55 @@ import (
...
 	"k8s.io/kubernetes/pkg/watch"
-	tapi "github.com/k8sdb/apimachinery/api"
 )
 
 type Controller struct {
 	*amc.Controller
+	// For Internal Cron Job
+	cron *cron.Cron
+	// Store Cron Job EntryID for further use
+	cronEntryIDs map[string]cron.EntryID
```
Comment (TamalSaha) [inline pkg/controller/controller.go]: Map are not safe for concurrent  operation in GOlang. Use https://github.com/orcaman/concurrent-map .

Comment (aerokite) [inline pkg/controller/controller.go]: used [concurrent-map](https://github.com/orcaman/concurrent-map)

File: pkg/controller/delete.go [hunk 2]
```diff
@@ -26,7 +27,8 @@ func (w *Controller) delete(elastic *tapi.Elastic) {
 	}
 
 	// Remove previous cron job if exist
-	if id, found := w.cronEntryIDs[elastic.Name]; found {
-		w.cron.Remove(id)
+	if id, found := w.cronEntryIDs.Get(elastic.Name); found {
```
Comment (TamalSaha) [inline pkg/controller/delete.go]: Use Pop(). https://github.com/orcaman/concurrent-map/blob/master/concurrent_map.go#L126

That will be atomic removal if present

File: pkg/controller/schedule.go [hunk 3]
```diff
@@ -88,8 +89,9 @@ func (b *backup) createDatabaseSnapshot() {
 // Backup schedule process with internal cron job.
 func (w *Controller) ScheduleBackup(elastic *tapi.Elastic) error {
 	// Remove previous cron job if exist
-	if id, found := w.cronEntryIDs[elastic.Name]; found {
-		w.cron.Remove(id)
+	if id, found := w.cronEntryIDs.Get(elastic.Name); found {
```
Comment (TamalSaha) [inline pkg/controller/schedule.go]: Use pop

File: pkg/controller/update.go [hunk 4]
```diff
@@ -47,8 +48,9 @@ func (w *Controller) update(oldElastic, updatedElastic *tapi.Elastic) {
 			}
 		} else {
 			// Remove previous cron job if exist
-			if id, found := w.cronEntryIDs[updatedElastic.Name]; found {
-				w.cron.Remove(id)
+			if id, found := w.cronEntryIDs.Get(updatedElastic.Name); found {
```
Comment (TamalSaha) [inline pkg/controller/update.go]: Use pop

File: pkg/controller/schedule.go (hunk 3 above)
Comment (aerokite) [inline pkg/controller/schedule.go]: fixed

File: pkg/controller/update.go (hunk 4 above)
Comment (aerokite) [inline pkg/controller/update.go]: fixed

File: pkg/controller/delete.go (hunk 2 above)
Comment (aerokite) [inline pkg/controller/delete.go]: fixed

File: pkg/controller/create.go [hunk 5]
```diff
@@ -19,6 +20,10 @@ const (
 	LabelDatabaseType          = "k8sdb.com/type"
 	LabelDatabaseName          = "elastic.k8sdb.com/name"
 	tagOperatorElasticsearch   = "0.1"
+	// Duration in Minute
+	// Check whether pod under StatefulSet is running or not
+	// Continue checking for this duration until failure
+	durationCheckStatefulSet = 30.0
```
Comment (TamalSaha) [inline pkg/controller/create.go]: Can this be a `time.Duration` ?

File: pkg/controller/create.go [hunk 6]
```diff
@@ -163,23 +183,56 @@ func (w *Controller) validateElastic(elastic *tapi.Elastic) bool {
...
 			return false
 		}
-		if storageClass == nil {
-			log.Errorln(fmt.Sprintf(`Spec.Storage.Class "%v" not found`, storage.Class))
+
+		if len(storage.AccessModes) == 0 {
+			storage.AccessModes = []kapi.PersistentVolumeAccessMode{
+				kapi.ReadWriteOnce,
+			}
+			log.Infoln(fmt.Sprintf(`Using "%v" as AccessModes in "%v"`, kapi.ReadWriteOnce, *storage))
```
Comment (TamalSaha) [inline pkg/controller/create.go]: We have `log.Infof()` . You will have to \n terminate yourself.

File: pkg/controller/lib.go [hunk 7]
```diff
@@ -5,12 +5,48 @@ import (
...
 	k8serr "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
 	kapps "k8s.io/kubernetes/pkg/apis/apps"
+	"k8s.io/kubernetes/pkg/apis/extensions"
 	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/util/intstr"
 )
 
+func (w *Controller) ensureElastic() {
+	resourceName := "elastic" + "." + tapi.V1beta1SchemeGroupVersion.Group
```
Comment (TamalSaha) [inline pkg/controller/lib.go]: Can we have a global constant tapi.kind = "elastic" ?

File: pkg/controller/create.go (hunk 5 above)
Comment (aerokite) [inline pkg/controller/create.go]: yes. Fixed

File: pkg/controller/lib.go (hunk 7 above)
Comment (aerokite) [inline pkg/controller/lib.go]: yes. Done

File: pkg/controller/create.go (hunk 6 above)
Comment (aerokite) [inline pkg/controller/create.go]: fixed

**Pull request merged**